contracts/StablecoinConverter.sol

```diff
@@ -108,4 +109,62 @@ contract StablecoinConverter is EpochTokenLocker {
     function tokenIdToAddressMap(uint16 id) public view returns (address) {
         return IdToAddressBiMap.getAddressAt(registeredTokens, id);
     }
-}
\ No newline at end of file
+
+    struct TradeData {
+        address owner;
+        uint volume;
+        uint16 orderId;
+    }
+
+    function submitSolution(
+        uint32 batchIndex,
+        address[] memory owner,  //tradeData is submitted as arrays
+        uint16[] memory orderId,
+        uint128[] memory volume,
+        uint128[] memory prices,  //list of prices for touched token only
+        uint16[] memory tokenIdForPrice  // price[i] is the price for the token with tokenID tokenIdForPrice[i]
+    ) public {
+        require(
+            batchIndex == getCurrentStateIndex() - 1,
+            "Solutions are no longer accepted for this batch"
+        );
+        uint len = owner.length;
+        for (uint i = 0; i < len; i++) {
+            Order memory order = orders[owner[i]][orderId[i]];
+            require(order.validFrom <= batchIndex, "Order is not yet valid");
+            require(order.validUntil >= batchIndex, "Order is no longer valid");
+            // Assume for now that we always have sellOrders
+            uint128 executedSellAmount = volume[i];
+            uint128 executedBuyAmount = uint128(
+                volume[i].mul(prices[findPriceIndex(order.buyToken, tokenIdForPrice)]) /
+                prices[findPriceIndex(order.sellToken, tokenIdForPrice)]
+            );
+            require(
+                executedSellAmount.mul(order.buyAmount) >= executedBuyAmount.mul(order.sellAmount),
+                "limit price not satisfied"
+            );
+            require(order.sellAmount >= executedSellAmount, "executedSellAmount bigger than specified in order");
+            orders[owner[i]][orderId[i]].buyAmount = uint128(
```

Comment:
`orders[owner[i]][orderId[i]]` was saved as memory, but here, we want to update the stored value in the smart contract.

Reply:
> owner[i], orderId[i] be refactored

agreed. I would like to refactor all variable names, once the solutionSubmission function is ready. The issue is that I have only 16 variables (because of stack deepness) and I am not yet sure, how to use them the best.

Reply:
Ok, I understand.

It's not usually a big deal, just this small things (comments & refactors) can helps reviewers to spot errors, but if you think you'll be moving them back and forth when you add new things is understandable.